Template addition

LaakkonenJussi opened the pull request.

LaakkonenJussi committed:
Implement template addition for replacing keys in profile files

Implement template addition to pass templates as key value pairs as cmd
line parameters to replace the keys in read profile file lines to allow
more customization and flexibility. This adds a new file called
template.c which contains all the functionality.

Motivation for this is to have the possibility to create profile files
where a specific key exists that can be customized per application that
is being run with firejail. For example, application name can be used in
a D-Bus name that is requested to be owned to avoid collision with other
applications requiring the same base name. This can be passed directly
then to firejail as a cmd line parameter when starting the application.

All templates are to be given via cmd line parameters in format:
 --template=KEY:VALUE

The keys and values are stored in a single linked list within template.c,
which is free'd when the keys in all read profile file (including
included profiles) lines have been replaced.

Each key can exist only once, existing hardcoded macros cannot be
overwritten. If any of these is violated firejail exits. A template key
cannot start with a digit and must contain alphanumeric chars only.

Each value must conform to following rules:
 - length is < 255 (D-Bus name length)
 - can contain alphabetical (a-zA-Z), integer (0-9) and '_-/.' chars but
   no '..'

In order to use the same DBUS_MAX_NAME_LENGTH it is moved from dbus.c to
firejail.h.

When processing the profile file lines the template keys are expected to
be written as other macros, ${TEMPLATE_KEY}. Template cannot be in the
beginning of the line. If the read line contains other internal macros
they are not replaced as they are processed later with more strict and
specific checks. It is known that using strtok_r() and doing the
tokenization in two steps, first by '$' and then by '{}' invalid
definitions such as ${{TEMPLATE_KEY2}} will pass the checks. The process
of replacing the keys can be described as follows to ease the
understanding of the code:

1. "whitelist ${HOME}/${key1}/path/to/${key2}.somewhere" tokens are:
    a: whitelist
    b: {HOME}/
    c: {key1}/path/to/
    d: {key2}.somewhere
2. Keys in the first token 'a' are ignored, it is the start of new str
3. Tokens 'b', 'c' and 'd' are passed to process_key_value
4. Each of the template keys are replaced with corresponding values, as
${HOME} is internal macro it is not replaced but added as is. Only the
first items in tokens, 'key1' and 'key2' are considered as proper keys to
have the values replaced, the remains are just added to the str.
5. Resulting string would be then:
   "whitelist ${HOME}/value1/path/to/value2.somewhere"

In order to avoid unnecessary duplication of each read profile line the
line is first checked to have at least one template key. If the template
key is not found firejail will exit with an error.

Man pages for firejail and firejail-profile are updated to include this
addition.

LaakkonenJussi commented on src/firejail/main.c:
```diff
@@ -2713,6 +2717,9 @@ int main(int argc, char **argv, char **envp) {
 		}
 	}
+
+	template_print_all();
```
Only for debug.

LaakkonenJussi replied:
I guess this is now obvious.

spiiroin commented on src/firejail/profile.c:
```diff
@@ -1706,16 +1706,40 @@ void profile_read(const char *fname) {
+
+		fprintf(stderr, "line %s\n", ptr_clean);
+
+		if (strrchr(ptr_clean, '$')) {
```
strRchr = odd choice

LaakkonenJussi replied:
Either a typo or thought error, changed to `strchr()` a while back.

spiiroin commented on src/firejail/profile.c:
```diff
@@ -1706,16 +1706,40 @@ void profile_read(const char *fname) {
-		if (*ptr == '#' || *ptr == '\0') {
-			free(ptr);
+		if (*ptr_clean == '#' || *ptr_clean == '\0') {
+			free(ptr_clean);
```
iside note: solating new variable "ptr_clean" within "has $" conditional would make the diff smaller and produce less "are leaks possible" brain load.

LaakkonenJussi replied:
Yeah well this is a wrong place anyways and need to check another approach on where to really replace the strings. It just dawned that this is done after the argument parser is executed.

LaakkonenJussi replied:
I think I changed that a lot last week, for the better.

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	if (!key || !*key || !value || !*value)
+		return NULL;
+
+	tmpl = (Template*)malloc(sizeof(Template));
```
Existing code does not cast mallocs - which is a good thing. Lets not start a bad habit...

LaakkonenJussi replied:
Old habit dies hard..

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	int key_len;
+
+	if (!tmpl_list) {
+		*err = -ENOENT;
```
Is this error reason detail really necessary? Existing code does not use this kind of pattern and even here in the new code that value seems to be ignored except for diagnostic logging.

LaakkonenJussi replied:
I think this is a remnant of code decay here. I had at first an approach to be more lib-like pass the list as parameter but I decided to hide all that functionality.

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+		iter = iter->next;
+	}
+
+	*err = -ENOKEY;
```
So, "item not in list" -> ENOKEY and "item not in list because list is empty" -> ENOENT. Is there some point to this differentiation?

LaakkonenJussi replied:
No, ENOKEY should be used.

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	key_len = strlen(key);
+	iter = tmpl_list;
+	while (iter) {
+		if (!strncmp(key, iter->key, key_len))
```
iter at "foobar" matches key "foo" - is that intentional?

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	}
+
+	if (template_get(tmpl->key, &err))
+		return -EALREADY;
```
So, this function may or may not take resource ownership -> that should be documented so that it is not easy to miss such detail.

spiiroin replied:
Also, matter of opinion etc: using if (tmpl_list && template_get(tmpl->key, &err)) -> would make the preceeding  "empty list special case" redundant.

LaakkonenJussi replied:
Idea here is to exit firejail if the same template is defined twice as cli arg. In such case EALREADY is reported back.

LaakkonenJussi replied:
I think this works as it is intended since for me it is quite routine to check the return value and free the content if adding is not possible due to duplicate content and thus, the newly allocated must be free'd by the caller.

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	Template *curr;
+
+	if (!tmpl_list)
+		return;
```
Matter of opinion, but this empty list special case could be removed without change in behavior -> just adds unnecessary complexity

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	if (!str || !*str)
+		return -EINVAL;
+
+	/* Keys or the values must start with an alphabetic char and must not
```
Values too? So we can't have $[FOOID} = 42?

LaakkonenJussi replied:
Well, yeah, I guess there is a need for plain ints.

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	fprintf(stdout, "Check %s list %p\n", arg, tmpl_list);
+
+	/* Only alphanumeric chars in template key. */
+	key = strtok((char*)arg, delim);
```
You can't claim to take "const char arg" and then pass it to strtok -> using cast to remove const is wrong and hides a potential bug.

spiiroin replied:
Also, strtok() is always a potential fubar situation to uninformed readed -> consider using strtok_r() even if you know that strtok would be just fine here.

LaakkonenJussi replied:
Hadn't yet gotten to cleanup this part ;)

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	key = strtok((char*)arg, delim);
+	if (is_valid_str(key, STR_CHECK_ALNUM) == -EINVAL) {
+		fprintf(stderr, "Error invalid template key \"%s\"", key);
+		exit(1);
```
So, we have a function that looks like a boolean predicate, but in reality returns error code - which is already confusing. Close inspection of is_valid_str() reveals that at the moment it returns either zero or -EINVAL -> this conditional here reads to me: all other errors are just fine... If the actual need here is boolean ok vs not-ok differentiation and these checks are not performed elsewhere, why bring those error codes in to picture at all?

LaakkonenJussi replied:
Thinking too far ahead and getting blind again? Using `bool` is neater, yes.

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+
+	/* Only a-zA-Z0-9_ */
+	value = strtok(NULL, delim);
+	if (is_valid_str(value, STR_CHECK_DBUS) == -EINVAL) {
```
btw, why are we limiting ourselves to dbus-clean strings only?

LaakkonenJussi replied:
Because the values can be used in the D-Bus names in profile/permissions as well, so I think they should follow the syntax there. But apparently that `dbus_check_name()` could be used instead.

spiiroin replied:
I'm more wondering about: if this is general mechanism that is not necessarily related only to dbus names and such, would requirement of dbus purity already at this level prohibit some otherwise useful ways to utilize this functionality? Would it be better / are there already checks in places where things are used in dbus related activity?

spiiroin commented on src/firejail/profile.c:
```diff
@@ -1706,16 +1706,40 @@ void profile_read(const char *fname) {
+		int err;
+		char *ptr;
+
+		fprintf(stderr, "line %s\n", ptr_clean);
```
I guess this and similar printf() calls are debug only -> a FIXME comment / something to signal it is not meant to stay might be nice to have.

LaakkonenJussi replied:
Yes, these are just for debugging. Too bad this does not have any debug-framework that could be used to make them more obvious/hidden in release builds.

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	Template *iter;
+
+	if (!tmpl_list)
+		return;
```
Again, this empty list special case does not accomplish too much.

LaakkonenJussi replied:
Changed a while back.

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	Template *tmpl;
+	const char *key;
+	const char *value;
+	char *delim = TEMPLATE_KEY_VALUE_DELIM;
```
'delim' can be const string

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+		break;
+	case -EALREADY:
+		fprintf(stderr, "Error template key \"%s\" already exists\n",
+								key);
```
Using set-once logic means: we can't have defaults in one config file and override them from another - why would we want to do that?

Or is the idea that these are defined only in command line?

LaakkonenJussi replied:
Yes these can be defined only via commandline using `--template=KEY:VALUE`.

LaakkonenJussi replied:
Which apparently still says `--macro` in main.c but need to change that as well.

LaakkonenJussi replied:
Changed, and apparently this is clear now.

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	}
+
+	tmpl = template_new(key, value);
+	err = template_add(tmpl);
```
random observation: This new+add pair builds assumed EINVAL reason that by the looks of it might lead to logically incorrect following events e.g. allocation failure would be reported as invalid key. Not that any of it is likely as both key and value have already been subjected to more strict checks than what template_new() uses and detectable allocation failures are unlikely too...

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+		return strdup(str2);
+
+	len = strlen(str2) + 1;
+	str1 = (char*)realloc(str1, strlen(str1) + len);
```
is C -> do not case realloc() return value

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	len = strlen(str2) + 1;
+	str1 = (char*)realloc(str1, strlen(str1) + len);
+
+	return strncat(str1, str2, len);
```
In nits and useless optimizations series: str1 can in theory be NULL.  Having that strlen(str1) above means we would not need to scan end of str1 again.

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	int i;
+
+	for (i = 0; internal_keys[i]; i++) {
+		if (!strncmp(key, internal_keys[i], strlen(internal_keys[i]))) {
```
strNcmp() again - is this leading part only check intentional?

LaakkonenJussi replied:
Overthinking and not thinking straight, replaced with plain `strcmp`

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+			goto err;
+
+		container = append_to_string(container, value);
+		if (!container)
```
that unchecked realloc() in append_to_string() plus variance in whether return value from append_to_string() is checked or not -> looks suspicious

LaakkonenJussi replied:
That is why this is WIP, things are just missing still.

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+		if (!str)
+			new_string = process_key_value(new_string, token, err);
+		else
+			new_string = append_to_string(new_string, token);
```
I'm confused about this "starting from second token" thing. What is that all about? [Checking that option/command itself is not a macro has already been checked earlier].

... also: '$' is removed from the 1st token when it looks like the intent might have been to pass it as-is and unmodified?

LaakkonenJussi replied:
Will add check that if there is a lone `$` it means an error. If that is any of the internal macros it will be replaced in the string, yeah, not the best way but since I resorted using `strtok_r()` I guess that is unavoidable unless that is to be done by hand, which may induce more errors..

See also https://github.com/LaakkonenJussi/firejail/pull/1/files#r630917495

LaakkonenJussi replied:
I'd suppose this is resolved but @spiiroin  do you still have something that needs to be clarified or do you think this needs fixing? I added comments to code and to commit msg about this as well.

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	if (new_string)
+		free(new_string);
+	else if (!err)
+		*err = -ENOMEM;
```
eh? if (err == NULL) *err = please_crash;

Also, going through all this passing error code pointers back and forth -> looks like it is uninitialized value originating from depths of profile_read() function -> it would not make sense for it to be 'if( *err ) *err = something:' either.

LaakkonenJussi replied:
oops a mistake there. Should be `else if (!*err)` thanks for pointing it out.

spiiroin replied:
... and that: what if *err is already non-zero due to originating from uninitialized-variable-in-stack like I think it is?

IMO you would either need to ensure that each and every error code is zero initialized, or not rely on previous content at all. In this case there is one place where reason is possibly ENOMEM and you could just set it unconditionally there and omit this implied ENOMEM bit?

spiiroin commented on src/firejail/profile.c:
```diff
@@ -1706,16 +1706,40 @@ void profile_read(const char *fname) {
+
+			if (ptr == NULL) {
+				fprintf(stderr, "Ignoring line \"%s\", contains "
+						"invalid template keys (err %d)\n",
```
Noting that error code value is not really used by the code itself: as places where those errors are triggered also contain error logging with details -> what are the chances that anybody reading the logs is also interested to know that all that human readable stuff is also known as "(err -127)".

LaakkonenJussi replied:
Maybe I should in general simplify that error use. Existing code does not use much different error codes but I guess here it could be good to report something that describes the actual cause. I did remove most of that to be as internal in `template.c` for reporting errors, might be enough?

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+	const char *value;
+
+	for (str = key; ; str = NULL) {
+		token = strtok_r(str, delim, &saveptr);
```
This splits at '{' or '}' - whichever comes first. Which would tokenize "foo  $bar}baz{" - how exactly?

LaakkonenJussi replied:
That is a bug that can be easily fixed.

spiiroin commented on src/firejail/template.c:
```diff
@@ -0,0 +1,429 @@
+			continue;
+		}
+
+		/* Only the first iteration is the template */
```
I think I lost the plot completely.

LaakkonenJussi replied:
Well since this tokenizes first by `$` there can be cases of: `include /path/to/${template1}/something/${template2}/somewhere` ->

1: tokens are a:`/path/to`,  b:`{template1}something/` and c:`{template2}/somewhere`
2: pass only a and b to this functon
3: Only the first item, in b:`template1` and c:`template2` are considered proper keys, the remains is just added to string.

I have an external test app for checking these cases. So in general when parsing the whole string in no circumstances the string should start in profile/permission file with a template.

spiiroin replied:
Ok, thanks for the clarification (this is "second level slicer" and you probably meant "2: pass only b and c to this function"?).

If needed, how does one have literal '$' / '{' chars in the final product? I have a feeling that eventually this kind of simple slicing loop will fail to handle something and becomes confusingly complex with special case handling or needs to be replaced with a bit more holistic decode+tokenize logic that allows escaping etc.

LaakkonenJussi replied:
Yes, `b` and `c` in that 2. My bad.

I have the `process_key_value()` implemented in a way that if there is an internal macro it will have its `${INTERNAL_MACRO}` format in the string. I'm eager to fix that macro.c to have simpler checker but lets see..

LaakkonenJussi replied:
And I'm not sure how complex this really needs to be. But anyways I haven't worked with Firejail for long so there may be something I've missed but from my point of view the lines in the profiles/permissions do seem to be quite simple.

spiiroin reviewed (commented):
Random observations about WIP code

LaakkonenJussi committed:
Add profile files to a list when processing arguments

Do not read the profile files when processing arguments. Instead prepend
them to a list that is processed after the arguments are processed in
order to be able to replace all template keys in the profile file lines.
The list is reversed before processing to have the profile files read in
the given order. Each item in the list is free'd after it has been
processed.
